File: gtci/17-modified-binary-search/python3/9-1-search-in-rotated-array-with-duplicates.py
class Solution:
    def search(self, arr, key):
        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (left + right) // 2

            if arr[mid] == key:
                return mid

            # duplicates, we don't know which side is sorted, so move left and right up by one
            if arr[left] == arr[mid] == arr[right]:
                left += 1
                right -= 1
                continue

            if arr[left] <= arr[mid]:  # left half is sorted
                if key >= arr[left] and key < arr[mid]:  # target in the left half
                    right = mid - 1
                else:
                    left = mid + 1  # target in the right half
            else:  # right half is sorted
                if key > arr[mid] and key <= arr[right]:
                    left = mid + 1  # target is in the right half
                else:
                    right = mid - 1  # target is in the left half

        return -1


if __name__ == "__main__":
    sol = Solution()

    assert sol.search([4, 5, 7, 9, 10, -1, 2], 10) == 4
    assert sol.search([4, 5, 7, 9, 10, -1, 2], -1) == 5
    assert sol.search([4, 5, 7, 9, 10, -1, 2], 6) == -1

    assert sol.search([3, 7, 3, 3, 3], 7) == 1
    assert sol.search([3, 3, 3, 3, 3], 7) == -1
    assert sol.search([3, 3, 3, 3, 3], 3) in [0, 1, 2, 3, 4]

    print("All test cases passed.")

File: gtci/17-modified-binary-search/cpp/9-1-search-in-rotated-array-with-duplicates.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
public:
  int search(const std::vector<int>& arr, int key) {
    if (arr.empty())
      return -1;
    
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
      const int mid = left + (right - left) / 2;

      if (arr[mid] == key)
        return mid;

      // duplicates, we don't know which side is sorted, so move left and right up by one
      if (arr[left] == arr[mid] && arr[right] == arr[mid]) {
        left++;
        right--;
        continue;
      }

      if (arr[left] <= arr[mid]) {
        // first half is sorted
        if (key >= arr[left] && key < arr[mid])   // target is in first half
          right = mid - 1;
        else
          left = mid + 1;                          // target is in second half
      } else {
        // second half is sorted
        if (key > arr[mid] && key <= arr[right])
          left = mid + 1;                          // target is in second half
        else
          right = mid - 1;                         // target is in first half
      }
    }

    return -1;
  }
};

int main() {
    Solution s;
    
    std::vector<int> arr1 {10, 15, 1, 3, 8};
    assert(s.search(arr1, 15) == 1);
    assert(s.search(arr1, 10) == 0);
    assert(s.search(arr1, 8) == 4);
    assert(s.search(arr1, 12) == -1);
    
    std::vector<int> arr2 {4, 5, 7, 9, 10, -1, 2};
    assert(s.search(arr2, 10) == 4);
    assert(s.search(arr2, -1) == 5);
    assert(s.search(arr2, 4) == 0);
    assert(s.search(arr2, 6) == -1);
    
    std::vector<int> arr3 {1};
    assert(s.search(arr3, 1) == 0);
    assert(s.search(arr3, 0) == -1);

    std::vector<int> arr4 {3, 7, 3, 3, 3};
    assert(s.search(arr4, 7) == 1);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
